README.md
[![npm version](https://badge.fury.io/js/@pahud%2Fcdktf-aws-eks.svg)](https://badge.fury.io/js/@pahud%2Fcdktf-aws-eks)
[![PyPI version](https://badge.fury.io/py/pahud-cdktf-aws-eks.svg)](https://badge.fury.io/py/pahud-cdktf-aws-eks)
[![release](https://github.com/pahud/cdktf-aws-eks/actions/workflows/release.yml/badge.svg)](https://github.com/pahud/cdktf-aws-eks/actions/workflows/release.yml)
[![construct hub](https://img.shields.io/badge/Construct%20Hub-available-blue)](https://constructs.dev/packages/@pahud/cdktf-aws-eks)

# cdktf-aws-eks

CDKTF construct library for Amazon EKS.

## Usage

The following sample creates:

1. A new VPC
1. Amazon EKS cluster(control plane)
2. The default nodegroup with the cluster
3. The 2nd nodegroup with spot instances


```ts
import { Cluster } from '@pahud/cdktf-aws-eks';

// create the cluster and the default nodegroup
const cluster = new Cluster(stack, 'demo-cluster', {
  version: KubernetesVersion.V1_21,
  scalingConfig: { minCapacity: 1 },
});

// create the optional 2nd nodegroup
cluster.addNodeGroup('NG2', {
  scalingConfig: {
    minCapacity: 1,
    maxCapacity: 10,
    desiredCapacity: 5,
  },
  capacityType: CapacityType.SPOT,
  instanceTypes: ['t3.large', 'c5.large', 'm5.large']
})
```

## Existing VPC subnets

To deploy in any existing VPC, specify the `privateSubnets` and `publicSubnets`(if any).

```ts
new Cluster(stack, 'demo-cluster', {
  privateSubnets: ['subnet-111','subnet-222','subnet-333' ],
  publicSubnets: ['subnet-444','subnet-555','subnet-666' ],
  version: KubernetesVersion.V1_21,
});
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.1](https://github.com/pahud/cdktf-aws-eks/compare/v0.1.1...v0.2.1) (2021-09-12)

## [0.2.0](https://github.com/pahud/cdktf-aws-eks/compare/v0.1.1...v0.2.0) (2021-09-12)

## [0.2.0](https://github.com/pahud/cdktf-aws-eks/compare/v0.1.0...v0.2.0) (2021-09-12)


test/default.test.ts
import { TerraformStack, Testing } from 'cdktf';
import { Cluster, KubernetesVersion } from '../src';

Testing.setupJest();
let app = Testing.app();
let stack = new TerraformStack(app, 'test');

describe('Unit testing using snapshots', () => {
  it('should match the snapshot', () => {
    expect(
      Testing.synthScope(() => {
        new Cluster(stack, 'Cluster', {
          version: KubernetesVersion.V1_21,
        });
      })).
      toMatchSnapshot();
  });
});

.projenrc.js
const { ConstructLibraryCdktf, DependenciesUpgradeMechanism, NpmAccess } = require('projen');

const AUTOMATION_TOKEN = 'PROJEN_GITHUB_TOKEN';

const project = new ConstructLibraryCdktf({
  author: 'Pahud Hsieh',
  authorAddress: 'pahudnet@gmail.com',
  cdktfVersion: '0.6.1',
  defaultReleaseBranch: 'main',
  name: '@pahud/cdktf-aws-eks',
  description: 'CDKTF construct library for Amazon EKS',
  repositoryUrl: 'https://github.com/pahud/cdktf-aws-eks',
  deps: [
    '@cdktf/provider-aws',
    '@cdktf/provider-kubernetes',
  ],
  peerDeps: [
    '@cdktf/provider-aws',
    '@cdktf/provider-kubernetes',
  ],
  depsUpgrade: DependenciesUpgradeMechanism.githubWorkflow({
    ignoreProjen: false,
    workflowOptions: {
      labels: ['auto-approve', 'auto-merge'],
      secret: AUTOMATION_TOKEN,
    },
  }),
  autoApproveOptions: {
    secret: 'GITHUB_TOKEN',
    allowedUsernames: ['pahud'],
  },
  npmAccess: NpmAccess.PUBLIC,
  publishToPypi: {
    distName: 'pahud-cdktf-aws-eks',
    module: 'pahud_cdktf_aws_eks',
  },
});

const common_exclude = ['cdktf.out', 'yarn-error.log', 'dependabot.yml', '.terraform', 'terraform.*'];
project.npmignore.exclude(...common_exclude, 'images', 'docs', 'website');
project.gitignore.exclude(...common_exclude);

project.synth();


src/cluster.ts
import { AwsProvider, EksCluster, DataAwsEksClusterAuth, IamRole, IamPolicyAttachment, DataAwsAvailabilityZones as AZ, DataAwsSubnetIds } from '@cdktf/provider-aws';
import * as k8s from '@cdktf/provider-kubernetes';
import { TerraformOutput, Token, ITerraformDependable } from 'cdktf';
import { Construct } from 'constructs';
import { NodeGroup, NodeGroupOptions, CapacityType, ScalingConfig } from '.';
import * as awsVpc from './imports/modules/terraform-aws-modules/vpc/aws';

/**
 * Properties for the Cluster
 */
export interface ClusterProps {
  /**
   * The AWS region to deploy.
   */
  readonly region?: string;
  /**
   * list of public subnetIds for an existing VPC.
   */
  readonly publicSubnets?: string[];
  /**
   * list of private subnetIds for an existing VPC.
   */
  readonly privateSubnets?: string[];
  /**
   * The Amazon EKS cluster name.
   */
  readonly clusterName?: string;
  /**
   * Kubernetes cluster version
   */
  readonly version: KubernetesVersion;
  /**
   * The scaling config of the default nodegroup.
   */
  readonly scalingConfig?: ScalingConfig;
  /**
   * capacity type of the nodegroup
   * @default CapacityType.ON_DEMAND
   */
  readonly capacityType?: CapacityType;
  /**
   * instance types of the default nodegroup.
   *
   * @default ['t3.large']
   */
  readonly instanceTypes?: string[];
}

/**
 * Kubernetes cluster version
 */
export class KubernetesVersion {
  /**
   * Kubernetes version 1.14
   */
  public static readonly V1_14 = KubernetesVersion.of('1.14');

  /**
   * Kubernetes version 1.15
   */
  public static readonly V1_15 = KubernetesVersion.of('1.15');

  /**
   * Kubernetes version 1.16
   */
  public static readonly V1_16 = KubernetesVersion.of('1.16');

  /**
   * Kubernetes version 1.17
   */
  public static readonly V1_17 = KubernetesVersion.of('1.17');

  /**
   * Kubernetes version 1.18
   */
  public static readonly V1_18 = KubernetesVersion.of('1.18');

  /**
   * Kubernetes version 1.19
   */
  public static readonly V1_19 = KubernetesVersion.of('1.19');

  /**
   * Kubernetes version 1.20
   */
  public static readonly V1_20 = KubernetesVersion.of('1.20');

  /**
   * Kubernetes version 1.21
   */
  public static readonly V1_21 = KubernetesVersion.of('1.21');

  /**
   * Custom cluster version
   * @param version custom version number
   */
  public static of(version: string) { return new KubernetesVersion(version); }
  /**
   *
   * @param version cluster version number
   */
  private constructor(public readonly version: string) { }
}


/**
 * The Amazon EKS Cluster with a default nodegroup
 */
export class Cluster extends Construct {
  readonly props: ClusterProps;
  readonly publicSubnets: string[];
  readonly privateSubnets: string[];
  readonly clusterName: string;
  readonly cluster: EksCluster;
  readonly vpc?: any;
  readonly vpcId?: string;
  readonly defaultNodeGroup?: NodeGroup;
  private readonly region: string;
  constructor(scope: Construct, id: string, props: ClusterProps) {
    super(scope, id);

    this.props = props;;
    this.region = props.region ?? 'us-east-1';
    new AwsProvider(this, 'aws', { region: this.region });

    // no private subnets given
    if (!props.privateSubnets) {
      const vpc = this._createVpc();
      this.vpc = vpc;
      this.vpcId = Token.asString(vpc.vpcIdOutput);
      this.privateSubnets = Token.asList(vpc.privateSubnetsOutput);
      this.publicSubnets = Token.asList(vpc.publicSubnetsOutput);
    } else {
      this.privateSubnets = props.privateSubnets!;
      this.publicSubnets = props.publicSubnets!;
    }

    // create the cluster
    this.clusterName = props.clusterName ?? `${id}cluster`;
    const cluster = new EksCluster(this, 'EksCluster', {
      name: this.clusterName,
      version: props.version.version,
      vpcConfig: [
        {
          // the cluster should associate with all available subnets
          subnetIds: this.vpcId ? this.getAllSubnetsFromVpcId(this.vpcId, [this.vpc]).ids :
            this.privateSubnets.concat(this.publicSubnets),
        },
      ],
      roleArn: this._createClusterRole().arn,
    });
    this.cluster = cluster;
    // cluster should be created after vpc
    if (this.vpc) {
      cluster.node.addDependency(this.vpc);
    }

    this.defaultNodeGroup = new NodeGroup(this, 'NG', {
      clusterName: cluster.name,
      subnets: this.privateSubnets,
      scalingConfig: props.scalingConfig,
      capacityType: props.capacityType,
      instanceTypes: props.instanceTypes,
      dependsOn: [this.cluster],
    });

    const clusterAuthData = new DataAwsEksClusterAuth(this, 'DataAwsEksClusterAuth', {
      name: this.clusterName,
    });

    let cert = `\${base64decode(${cluster.certificateAuthority})}`;

    const k8sprovider = new k8s.KubernetesProvider(this, 'Kubernetes', {
      host: Token.asString(cluster.endpoint),
      token: Token.asString(clusterAuthData.token),
    });
    k8sprovider.addOverride('cluster_ca_certificate', cert);

    new TerraformOutput(this, 'KubelctlCommand', {
      value: `aws --region ${this.region} eks update-kubeconfig --name ${this.clusterName}`,
    });

  }
  private _createVpc() {
    const vpc = new awsVpc.TerraformAwsModulesVpcAws(this, 'Vpc', {
      cidr: '10.0.0.0/16',
      azs: new AZ(this, 'AZs', {
        state: 'available',
      }).names,
      publicSubnets: ['10.0.1.0/24', '10.0.2.0/24', '10.0.3.0/24'],
      privateSubnets: ['10.0.11.0/24', '10.0.12.0/24', '10.0.13.0/24'],
      singleNatGateway: true,
      enableNatGateway: true,
      oneNatGatewayPerAz: false,
    });
    return vpc;
  }

  private _createClusterRole(): IamRole {
    const role = new IamRole(this, 'ClusterRole', {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Sid: '',
            Principal: {
              Service: 'eks.amazonaws.com',
            },
          },
        ],
      }),
    });
    new IamPolicyAttachment(this, 'AmazonEKSClusterPolicyAttachment', {
      name: 'AmazonEKSClusterPolicyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKSClusterPolicy',
      roles: [role.name],
    });
    new IamPolicyAttachment(this, 'AmazonEKSVPCResourceControllerAttachment', {
      name: 'AmazonEKSVPCResourceControllerAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKSVPCResourceController',
      roles: [role.name],
    });
    return role;
  }
  private getAllSubnetsFromVpcId(vpcId: string, dependable?: ITerraformDependable[]) {
    return new DataAwsSubnetIds(this, `${vpcId}subnets`, {
      vpcId,
      dependsOn: dependable,
    });
  }
  public addNodeGroup(id: string, options: NodeGroupOptions) {
    new NodeGroup(this, id, {
      ...options,
      clusterName: this.clusterName,
      nodeRole: options.nodeRole ?? this.defaultNodeGroup?.nodeGroupRoleArn,
      subnets: options.subnets ?? this.privateSubnets,
      dependsOn: [this.cluster],
    });
  }
}


src/nodegroup.ts

import { EksNodeGroup, IamRole, IamPolicyAttachment } from '@cdktf/provider-aws';
import { ITerraformDependable } from 'cdktf';
import { Construct } from 'constructs';


export interface ScalingConfig {
  readonly desiredCapacity?: number;
  readonly minCapacity?: number;
  readonly maxCapacity?: number;
}

export enum CapacityType {
  SPOT = 'SPOT',
  ON_DEMAND = 'ON_DEMAND'
}

export interface NodeGroupBaseOptions {
  /**
     * nodegroup role arn
     * @default - The IAM role for the default nodegroup.
     */
  readonly nodeRole?: string;
  /**
     * scaling configuration for the nodegroup.
     * @see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_node_group#scaling_config-configuration-block
     */
  readonly scalingConfig?: ScalingConfig;
  /**
     * capacity type of the nodegroup.
     * @default CapacityType.ON_DEMAND
     */
  readonly capacityType?: CapacityType;
  /**
     * instance types of the nodegroup.
     * @default ['t3.large']
     */
  readonly instanceTypes?: string[];
  /**
     * resources to depend on;
     */
  readonly dependsOn?: ITerraformDependable[];
}

export interface NodeGroupOptions extends NodeGroupBaseOptions {
  /**
     * subnet IDs for the nodegroup.
     */
  readonly subnets?: string[];
}

export interface NodeGroupProps extends NodeGroupBaseOptions {
  /**
     * cluster name
     */
  readonly clusterName: string;
  /**
     * subnet IDs for the nodegroup.
     */
  readonly subnets: string[];
}

export class NodeGroup extends Construct {
  readonly nodeGroupRoleArn: string;
  private readonly desiredCapacity: number;
  private readonly minCapacity: number;
  private readonly maxCapacity: number;
  constructor(scope: Construct, id: string, props: NodeGroupProps) {
    super(scope, id);
    this.minCapacity = props.scalingConfig?.minCapacity ?? 0;
    this.desiredCapacity = props.scalingConfig?.desiredCapacity ?? this.minCapacity;
    this.maxCapacity = props.scalingConfig?.maxCapacity ?? (this.desiredCapacity > 0) ?
      this.desiredCapacity : 1;

    this.nodeGroupRoleArn = props.nodeRole ?? this._createNodeGroupRole().arn;

    new EksNodeGroup(this, 'NodeGroup', {
      clusterName: props.clusterName,
      nodeRoleArn: this.nodeGroupRoleArn,
      subnetIds: props.subnets,
      scalingConfig: [
        {
          desiredSize: this.desiredCapacity,
          minSize: this.minCapacity,
          maxSize: this.maxCapacity,
        },
      ],
      capacityType: props.capacityType ?? CapacityType.ON_DEMAND,
      instanceTypes: props.instanceTypes ?? ['t3.large'],
      dependsOn: props.dependsOn,
    });
  }
  private _createNodeGroupRole(): IamRole {
    const role = new IamRole(this, 'MNGRole', {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Sid: '',
            Principal: {
              Service: 'ec2.amazonaws.com',
            },
          },
        ],
      }),
    });
    new IamPolicyAttachment(this, 'AmazonEKSWorkerNodePolicyAttachment', {
      name: 'AmazonEKSWorkerNodePolicyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy',
      roles: [role.name],
    });
    new IamPolicyAttachment(this, 'AmazonEKS_CNI_PolicyAttachment', {
      name: 'AmazonEKS_CNI_PolicyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy',
      roles: [role.name],
    });
    new IamPolicyAttachment(this, 'AmazonEC2ContainerRegistryReadOnlyAttachment', {
      name: 'AmazonEC2ContainerRegistryReadOnlyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly',
      roles: [role.name],
    });
    return role;
  }
}



src/main.ts
import { AwsProvider, EksNodeGroup, EksCluster, DataAwsEksClusterAuth, IamRole, IamPolicyAttachment, DataAwsAvailabilityZones as AZ, DataAwsSubnetIds } from '@cdktf/provider-aws';
import * as k8s from '@cdktf/provider-kubernetes';
import { TerraformOutput, Token, ITerraformDependable } from 'cdktf';
import { Construct } from 'constructs';
import * as awsVpc from './imports/modules/terraform-aws-modules/vpc/aws';

/**
 * Properties for the Cluster
 */
export interface ClusterProps {
  /**
   * The AWS region to deploy.
   */
  readonly region?: string;
  /**
   * list of public subnetIds for an existing VPC.
   */
  readonly publicSubnets?: string[];
  /**
   * list of private subnetIds for an existing VPC.
   */
  readonly privateSubnets?: string[];
  /**
   * The Amazon EKS cluster name.
   */
  readonly clusterName?: string;
  /**
   * Kubernetes cluster version
   */
  readonly version: KubernetesVersion;
  /**
   * The desired capacity for the nodegroup.
   * @default - minCapacity
   */
  readonly desiredCapacity?: number;
  /**
   * min capacity for the nodegroup
   * @default 0
   */
  readonly minCapacity?: number;
  /**
   * max capacity for the nodegroup
   * @default - desiredCapacity
   */
  readonly maxCapacity?: number;
  /**
   * capacity type of the nodegroup
   * @default CapacityType.ON_DEMAND
   */
  readonly capacityType?: CapacityType;
  /**
   * instance types of the default nodegroup.
   *
   * @default ['t3.large']
   */
  readonly instanceTypes?: string[];
}

export enum CapacityType {
  SPOT = 'SPOT',
  ON_DEMAND = 'ON_DEMAND'
}

/**
 * Kubernetes cluster version
 */
export class KubernetesVersion {
  /**
   * Kubernetes version 1.14
   */
  public static readonly V1_14 = KubernetesVersion.of('1.14');

  /**
   * Kubernetes version 1.15
   */
  public static readonly V1_15 = KubernetesVersion.of('1.15');

  /**
   * Kubernetes version 1.16
   */
  public static readonly V1_16 = KubernetesVersion.of('1.16');

  /**
   * Kubernetes version 1.17
   */
  public static readonly V1_17 = KubernetesVersion.of('1.17');

  /**
   * Kubernetes version 1.18
   */
  public static readonly V1_18 = KubernetesVersion.of('1.18');

  /**
   * Kubernetes version 1.19
   */
  public static readonly V1_19 = KubernetesVersion.of('1.19');

  /**
   * Kubernetes version 1.20
   */
  public static readonly V1_20 = KubernetesVersion.of('1.20');

  /**
   * Kubernetes version 1.21
   */
  public static readonly V1_21 = KubernetesVersion.of('1.21');

  /**
   * Custom cluster version
   * @param version custom version number
   */
  public static of(version: string) { return new KubernetesVersion(version); }
  /**
   *
   * @param version cluster version number
   */
  private constructor(public readonly version: string) { }
}


/**
 * The Amazon EKS Cluster with a default nodegroup
 */
export class Cluster extends Construct {
  readonly props: ClusterProps;
  readonly publicSubnets: string[];
  readonly privateSubnets: string[];
  readonly clusterName: string;
  readonly vpc?: any;
  readonly vpcId?: string;
  private readonly desiredCapacity: number;
  private readonly minCapacity: number;
  private readonly maxCapacity: number;
  constructor(scope: Construct, id: string, props: ClusterProps) {
    super(scope, id);

    this.props = props;
    this.minCapacity = props.minCapacity ?? 0;
    this.desiredCapacity = props.desiredCapacity ?? this.minCapacity;
    this.maxCapacity = props.maxCapacity ?? this.desiredCapacity;
    new AwsProvider(this, 'aws', {
      region: props.region ?? 'us-east-1',
    });

    // no private subnets given
    if (!props.privateSubnets) {
      const vpc = this._createVpc();
      this.vpc = vpc;
      this.vpcId = Token.asString(vpc.vpcIdOutput);
      this.privateSubnets = Token.asList(vpc.privateSubnetsOutput);
      this.publicSubnets = Token.asList(vpc.publicSubnetsOutput);
    } else {
      this.privateSubnets = props.privateSubnets!;
      this.publicSubnets = props.publicSubnets!;
    }

    // create the cluster
    this.clusterName = props.clusterName ?? `${id}cluster`;
    const cluster = new EksCluster(this, 'EksCluster', {
      name: this.clusterName,
      version: props.version.version,
      vpcConfig: [
        {
          // the cluster should associate with all available subnets
          subnetIds: this.vpcId ? this.getAllSubnetsFromVpcId(this.vpcId, [this.vpc]).ids :
            this.privateSubnets.concat(this.publicSubnets),
        },
      ],
      roleArn: this._createClusterRole().arn,
    });

    // cluster should be created after vpc
    if (this.vpc) {
      cluster.constructNode.addDependency(this.vpc);
    }

    new EksNodeGroup(this, 'NG', {
      clusterName: cluster.name, // ensure the dependency
      nodeRoleArn: this._createNodeGroupRole().arn,
      subnetIds: this.privateSubnets,
      scalingConfig: [
        {
          desiredSize: this.desiredCapacity,
          minSize: this.minCapacity,
          maxSize: this.maxCapacity,
        },
      ],
      capacityType: props.capacityType ?? CapacityType.ON_DEMAND,
      instanceTypes: props.instanceTypes ?? ['t3.large'],
    });

    const clusterAuthData = new DataAwsEksClusterAuth(this, 'DataAwsEksClusterAuth', {
      name: this.clusterName,
    });

    let cert = `\${base64decode(${cluster.certificateAuthority})}`;

    const k8sprovider = new k8s.KubernetesProvider(this, 'Kubernetes', {
      host: Token.asString(cluster.endpoint),
      token: Token.asString(clusterAuthData.token),
    });
    k8sprovider.addOverride('cluster_ca_certificate', cert);

    new TerraformOutput(this, 'KubelctlCommand', {
      value: `aws eks update-kubeconfig --name ${this.clusterName}`,
    });

  }
  private _createVpc() {
    const vpc = new awsVpc.TerraformAwsModulesVpcAws(this, 'Vpc', {
      cidr: '10.0.0.0/16',
      azs: new AZ(this, 'AZs', {
        state: 'available',
      }).names,
      publicSubnets: ['10.0.1.0/24', '10.0.2.0/24', '10.0.3.0/24'],
      privateSubnets: ['10.0.11.0/24', '10.0.12.0/24', '10.0.13.0/24'],
      singleNatGateway: true,
      enableNatGateway: true,
      oneNatGatewayPerAz: false,
    });
    return vpc;
  }
  private _createNodeGroupRole(): IamRole {
    const role = new IamRole(this, 'MNGRole', {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Sid: '',
            Principal: {
              Service: 'ec2.amazonaws.com',
            },
          },
        ],
      }),
    });
    new IamPolicyAttachment(this, 'AmazonEKSWorkerNodePolicyAttachment', {
      name: 'AmazonEKSWorkerNodePolicyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy',
      roles: [role.name],
    });
    new IamPolicyAttachment(this, 'AmazonEKS_CNI_PolicyAttachment', {
      name: 'AmazonEKS_CNI_PolicyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy',
      roles: [role.name],
    });
    new IamPolicyAttachment(this, 'AmazonEC2ContainerRegistryReadOnlyAttachment', {
      name: 'AmazonEC2ContainerRegistryReadOnlyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly',
      roles: [role.name],
    });
    return role;
  }
  private _createClusterRole(): IamRole {
    const role = new IamRole(this, 'ClusterRole', {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Sid: '',
            Principal: {
              Service: 'eks.amazonaws.com',
            },
          },
        ],
      }),
    });
    new IamPolicyAttachment(this, 'AmazonEKSClusterPolicyAttachment', {
      name: 'AmazonEKSClusterPolicyAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKSClusterPolicy',
      roles: [role.name],
    });
    new IamPolicyAttachment(this, 'AmazonEKSVPCResourceControllerAttachment', {
      name: 'AmazonEKSVPCResourceControllerAttachment',
      policyArn: 'arn:aws:iam::aws:policy/AmazonEKSVPCResourceController',
      roles: [role.name],
    });
    return role;
  }
  private getAllSubnetsFromVpcId(vpcId: string, dependable?: ITerraformDependable[]) {
    return new DataAwsSubnetIds(this, `${vpcId}subnets`, {
      vpcId,
      dependsOn: dependable,
    });
  }
}
